modules/compiler/spirv-ll/source/builder_debug_info.cpp

```diff
@@ -15,29 +15,2444 @@
+
+  size_t getScopeIdx() const { return 6 + dummy_offset; }
+
+ private:
+  size_t dummy_offset = 0;
+};
+
+template <>
+llvm::Error DebugInfoBuilder::create<DebugImportedEntity>(
+    OpExtInst const &opc) {
+  auto *op = module.create<DebugImportedEntity>(opc);
+  uint32_t line = op->Line();
+
+  auto entity_or_error = translateDebugInst<llvm::DINode>(op->Entity());
+  if (auto err = entity_or_error.takeError()) {
+    return err;
+  }
+  llvm::DINode *const entity = entity_or_error.get();
+
+  llvm::DIFile *file = nullptr;
+  if (op->Source()) {
+    auto file_or_error = translateDebugInst<llvm::DIFile>(op->Source());
+    if (auto err = file_or_error.takeError()) {
+      return err;
+    }
+    file = file_or_error.get();
+  }
+  // If we haven't a file, we can't have a non-zero line number. LLVM asserts on
+  // this.
+  if (!file) {
+    line = 0;
+  }
+
+  llvm::DIScope *scope = nullptr;
+  if (op->Scope()) {
+    auto scope_or_error = translateDebugInst<llvm::DIScope>(op->Scope());
+    if (auto err = scope_or_error.takeError()) {
+      return err;
+    }
+    scope = scope_or_error.get();
+  }
+
+  llvm::DIBuilder &dib = getDIBuilder(op);
+
+  if (op->Tag() == OpenCLDebugInfo100ImportedModule) {
+    if (!entity) {
+      dib.createImportedModule(
+          scope, static_cast<llvm::DIImportedEntity *>(nullptr), file, line);
+      return llvm::Error::success();
+    }
+    if (auto *const di_module = llvm::dyn_cast<llvm::DIModule>(entity)) {
+      dib.createImportedModule(scope, di_module, file, line);
+      return llvm::Error::success();
+    }
+    if (auto *const di_ie = llvm::dyn_cast<llvm::DIImportedEntity>(entity)) {
+      dib.createImportedModule(scope, di_ie, file, line);
+      return llvm::Error::success();
+    }
+    if (auto *const di_namespace = llvm::dyn_cast<llvm::DINamespace>(entity)) {
+      dib.createImportedModule(scope, di_namespace, file, line);
+      return llvm::Error::success();
+    }
+    return makeStringError("Unhandled imported module");
+  }
+
+  if (op->Tag() == OpenCLDebugInfo100ImportedDeclaration) {
+    std::optional<std::string> name = module.getDebugString(op->Name());
+    if (!name) {
+      return makeStringError(
+          "Could not find OpString 'Name' for DebugImportedEntity " +
+          getIDAsStr(op->IdResult(), &module));
+    }
+    if (auto *const di_glob =
+            llvm::dyn_cast_if_present<llvm::DIGlobalVariableExpression>(
+                entity)) {
+      dib.createImportedDeclaration(scope, di_glob->getVariable(), file, line,
+                                    *name);
+    } else {
+      dib.createImportedDeclaration(scope, entity, file, line, *name);
+    }
+    return llvm::Error::success();
+  }
+
+  return makeStringError("Unexpected imported entity kind");
+}
+
+llvm::Expected<llvm::MDNode *>
+DebugInfoBuilder::translateTemplateTemplateParameterOrTemplateParameterPack(
+    const OpExtInst *op) {
```

Comment:
Yeah I was really hopeful about that, but unfortunately it could also be `DebugInfoNone`. I believe `Template Name` can also be `DebugInfoNone` (not sure - it's not immediately clear if that makes sense to "optimize out").

I think seeing a positive `DebugSource` match for operand 2 is a sure-fire sign we have `DebugTypeTemplateParameterPack`.

I think seeing a positive `OpString` match for operand 2 is a sure-fire sign we have `DebugTypeTemplateTemplateParameter`.

If neither of these are true (which does happen, sadly), we necessarily need to check operands which may be  `spv::Id` or may be a literal number, depending on which instruction we are inspecting. Either 3 or 5 are interesting.

Fundamentally we don't ever really _know_ if an operand is a `spv::Id` or a literal number. So I think there'll always be a bit of ambiguity if we check operand 3 or 5. That's why I tried to go with matching pairs of operands 3 and 5, to hopefully mitigate the risk that we're just looking at numbers which happen to correspond to valid IDs.

This is the "algorithm" I tried to encode in this function. Maybe it's hard to follow?

Reply:
Ah, right. Let me write this down:

* Operand 0: always `OpTypeVoid`, which tells us nothing.
* Operand 1: always `OpString`, which tells us nothing.
* Operand 2: either `OpString` (`DebugTypeTemplateTemplateParameter`) or `DebugSource` (`DebugTypeTemplateParameterPack`). or `DebugInfoNone`.
* Operand 3: either `DebugSource` (`DebugTypeTemplateTemplateParameter`) or `word` (`DebugTypeTemplateParameterPack`), which cannot be distinguished reliably.
* Operand 4: always `word`, which tells us nothing.
* Operand 5: either `word` (`DebugTypeTemplateTemplateParameter`) or `DebugTypeTemplateParameter` (`DebugTypeTemplateParameterPack`), which cannot be distinguished reliably.

So for operand 3 and 5, if it cannot be a valid ID, it must be a `word`, which tells us what instruction we're looking at. But it's possible that both are a valid ID, in which case we can't figure anything out. Thanks, that makes sense, and I think that's the best that we can do.